### ISSUE TITLE
Fix ENOENT error for Jest test in function-no-unknown

### ISSUE THREAD

**Issue by 100terres** [stylelint/stylelint#5904]
### What steps are needed to reproduce the bug?

I've documented the issue on a library used by `stylelint`

- https://github.com/niksy/css-functions-list/issues/2#issue-1133605485

- Use the stylelint with the node API
- Using `stylelint-config-recommended@v7.0.0`
- It's running in a `jest@v26.6.3` test
- Using `node@v14.18.2`
- EDIT: In a `TypeScript@4.5.5` / babel environment

When running this the `function-no-unknown` rule throws this error

```
ENOENT: no such file or directory, open '/index.json'
```

Here is where it fails https://github.com/react-forked/dnd/blob/e18cfa53c3f6bb8f085137c934a4a0d9b8332671/test/unit/view/style-marshal/get-styles.spec.ts

Here is the error in CircleCI
https://app.circleci.com/pipelines/github/react-forked/dnd/229/workflows/9d30597b-ac76-4548-b2cb-3416ab0a03fe/jobs/1331

### What Stylelint configuration is needed to reproduce the bug?

```json
{
  "extends": ["stylelint-config-recommended"]
}
```

### How did you run Stylelint?

Node `stylelint.lint(...)`

### Which version of Stylelint are you using?

14.5.0

### What did you expect to happen?

To not throw an error

### What actually happened?

It throws this error 

```
ENOENT: no such file or directory, open '/index.json'
```

### Does the bug relate to non-standard syntax?

Nop

### Proposal to fix the bug

As suggested [here](https://github.com/niksy/css-functions-list/issues/2#issue-1133605485) I've made it work by using `path.resolve` [here](https://github.com/niksy/css-functions-list/blob/master/index.js#L6) in this lib and directly using the return of the `require("css-functions-list")` [here](https://github.com/stylelint/stylelint/blob/57bc6ef3917943ed740223055dea9030b142ebc0/lib/rules/function-no-unknown/index.js#L48) without using the `new URL`.

**Comment by 100terres** [comment]
I've look around a bit and I found that it might also be related to `TypeScript` (I forgot to mention that I was using it).

- https://github.com/stylelint/stylelint/issues/5291#issuecomment-963830714
- https://github.com/microsoft/TypeScript/issues/46452
- https://github.com/facebook/jest/issues/9430

**Comment by ybiquitous** [comment]
The `require("css-functions-list")` way had a problem on Windows. See below:

- https://github.com/stylelint/stylelint/pull/5865#issuecomment-1022136119
- https://github.com/stylelint/stylelint/pull/5865/commits/22fe22e2f99d2490295eaa31c36f57ee0deba829

**Comment by ybiquitous** [comment]
@100terres I've open a pull request that may fix your problem. Can you try it, please?

```diff
-    "stylelint": "^14.5.0",
+    "stylelint": "github:stylelint/stylelint#issue-5904",
```

**Comment by ybiquitous** [comment]
Sorry, PR #5906 doesn't work. 😢 
See https://github.com/stylelint/stylelint/pull/5906#issuecomment-1037651644

**Comment by ybiquitous** [comment]
@100terres I've opened PR #5916 that no longer uses `URL` thanks to `css-functions-list@3.0.1`.
Can you try whether to resolve `ENOENT`, please?

```diff
-    "stylelint": "^14.5.0",
+    "stylelint": "github:stylelint/stylelint#bump-css-functions-list",
```

**Comment by 100terres** [comment]
Hi @ybiquitous thank you for your time! :slightly_smiling_face: 

I've just tested it [here](https://github.com/react-forked/dnd/commit/dc933f262488ed2ad372125c92ff91643cea2a62) and know it's giving me this error. I didn't had the time to investigate if I need to change something in my config (with `jest` or `TypeScript`)

```
TypeError: The URL must be of scheme fileTypeError [ERR_INVALID_URL_SCHEME]: The URL must be of scheme file
```

Here is the result of the CI https://app.circleci.com/pipelines/github/react-forked/dnd/232/workflows/bf31edb7-16e2-4c43-9835-0271ab13ef75/jobs/1349

And now event if I [disable the `function-no-unknown` rule](https://github.com/react-forked/dnd/commit/b16f77553fc53ac4fa55a5a4fd64e36851831400) I still get the error described above, but only for the first test that run. [Circle CI](https://app.circleci.com/jobs/github/react-forked/dnd/1355).

**Comment by ybiquitous** [comment]
@100terres Thanks for the detailed reproduction. I'm not sure why `ERR_INVALID_URL_SCHEME` is raised... 🤷🏼

**Comment by ybiquitous** [comment]
@100terres The problem seems due to `jsdom`. When changing `testEnvironment` to `'node'`, it works:

```diff
diff --git a/jest.config.js b/jest.config.js
index 08a90c82..52634cf4 100644
--- a/jest.config.js
+++ b/jest.config.js
@@ -12,4 +12,5 @@ module.exports = {
     'jest-watch-typeahead/testname',
   ],
   verbose: true,
+  testEnvironment: 'node',
 };
```

```console
$ npx jest test/unit/view/style-marshal/get-styles.spec.ts
 PASS  test/unit/view/style-marshal/get-styles.spec.ts
  ✓ should generate valid always styles (117 ms)
  ✓ should generate valid resting styles (5 ms)
  ✓ should generate valid dragging styles (6 ms)
  ✓ should generate valid dropAnimating styles (3 ms)
  ✓ should generate valid userCancel styles (1 ms)

Test Suites: 1 passed, 1 total
Tests:       5 passed, 5 total
Snapshots:   0 total
Time:        0.946 s, estimated 2 s
Ran all test suites matching /test\/unit\/view\/style-marshal\/get-styles.spec.ts/i.
```

**Comment by ybiquitous** [comment]
`css-functions-list@3.0.1` includes the following code on CJS:

`node_modules/css-functions-list/cjs/index.js`:

```js
'use strict';

var url = require('url');

/**
 * Path to CSS functions list JSON file.
 */
const location = url.fileURLToPath(new url.URL('index.json', (typeof document === 'undefined' ? new (require('u' + 'rl').URL)('file:' + __filename).href : (document.currentScript && document.currentScript.src || new URL('index.js', document.baseURI).href))));

module.exports = location;
//# sourceMappingURL=index.js.map
```

I try embedding the debug code into the above file:

```diff
 var url = require('url');
+console.log('document:', document)
+console.log('document.currentScript:', document.currentScript)
+console.log('document.baseURI:', document.baseURI)
 /**
  * Path to CSS functions list JSON file.
  */
```

Then, run the failed test:

```console
$ npx jest test/unit/view/style-marshal/get-styles.spec.ts
  console.log
    document: Document {
      location: [Getter/Setter],
      [Symbol(DOM SymbolTree)]: SymbolTreeNode {
        parent: null,
        previousSibling: null,
        nextSibling: null,
        firstChild: null,
        lastChild: null,
        childrenVersion: 0,
        childIndexCachedUpTo: null,
        cachedIndex: -1,
        cachedIndexVersion: NaN
      }
    }

      at Object.<anonymous> (node_modules/css-functions-list/cjs/index.js:4:9)

  console.log
    document.currentScript: null

      at Object.<anonymous> (node_modules/css-functions-list/index.js:3:1)

  console.log
    document.baseURI: http://localhost/

      at Object.<anonymous> (node_modules/css-functions-list/index.js:4:1)

 FAIL  test/unit/view/style-marshal/get-styles.spec.ts
...
```

Unexpectedly, `document.baseURI` returns `http://localhost`, so `new URL('index.js', document.baseURI)` doesn't point to a local file path in the `node_modules/css-functions-list` directory.

**Comment by 100terres** [comment]
Hi @ybiquitous thank you a lot for taking the time to investigate this futher! :slightly_smiling_face:

I've just tested it and has [you mention](https://github.com/stylelint/stylelint/issues/5904#issuecomment-1039704057) the test works in the `node` environment.

It feels like that inside a `jest` test using `stylelint` should be running in a `node` environment and not in a `jsdom` environment (which mimic a browser environment).

**Comment by 100terres** [comment]
Also JSDOM does not seems to support `esm` yet https://github.com/jsdom/jsdom/issues/2475

**Comment by ybiquitous** [comment]
@100terres Thanks for sharing the knowledge. 👍🏼
